Hardening: Moving to `Owned/Borrowed` everywhere.

dkorolev opened the pull request:
### NOTE: This pull request is into `C5T/Current@alt-master`, not into `master`. The change is too dramatic to just silently make it into the master branch. Details on our next sync-up.

* * *

# Refactoring TL;DR

* `Owned<>`/`Borrowed<>` everywhere.
* Safer HTTP handlers registration.
* Refactored stream publishing ownership.
* Refactored storage-to-stream ownership.
* Refactored master-follower flips (TODO(dkorolev): Finish this, start from Sherlock/Stream).
* Karl tests are no longer flaky.
* A bunch of small tweaks.

# Implementation details

## `Owned<>`/`Borrowed<>`

I've used `Owned<>`/`Borrowed<>` in several places, most notably HTTP route registration / deregistration / serving, and stream publisher exposure.

The default `Borrowed<>` is now callback-free; it can even be copy-constructed. A special type `BorrowedWithCallback<>` has been introduced to deal with a borrowable object which may need to be woken up in order to be destroyed.

`Owned<>`/`Borrowed<>` no longer throw, ever. Even if the `Owner<>` is in the "being destructed, blocked until all the borrowers are gone" mode, the users of borrowed objects can do whatever they want, including dereferencing those object and spawning new borrowed ones. I concluded this design would be better than forcing the user to always keep in mind that any `Owned<>/Borrowed<>` action can possibly throw.

I've had debugging troubles with `ARGS...`-proxying constructors of `Owned<>`, so I disabled them :-) Well, kind of. They are still enabled outside Current, while within our own code I've made sure to always use `current::MakeOwned<T>`, with a fully specified `T`, much like `std::make_{shared,unique}`. After some deliberation, I'm convinced it's the cleanest we can go about it, while still maintaining the overall "debuggability".

The common base for `Owned` and `Borrowed` is be called `WeakBorrowed<T>` (ex. `Borrowable<T>`). TL;DR: if an object X exposes `x.BorrowY()` (returning `Borrowed<Y>`), it may also expose `x.Y()` (generally speaking, two versions, const and mutable), returning `WeakBorrowed<Y>`. The former syntax engages in the "smart reference counting" logic while the latter does not. Needless to say, when using `WeakBorrowed<Y>`  the user of `x.Y()` assumes full responsibility for the "being temporarily" used Y to not go out of (master) scope in the meantime. The most generic usecase of `.Y()` for some `y_` is if `y_` is a data class member, and `Y()` is used as a one-liner construct such as `stream->Data()->Size()` (whereas `stream->BorrowData()->Size()`, while doing the same thing, would be an order of magnitude less efficient because of the reference counting, mutexes, etc.)

## Stream 

This is another large change. Each stream now has a "detachable" `Publisher`, and it's stored as the combination of three members: a mutex to lock the next two, an `Optional<Owned<Publisher>> owned_publlisher_`, and an `Optional<Borrowed<Publisher>> borrowable_publisher_`. Here's the trick:

1. When an external-to-stream user wants to borrow a publisher, it uses `borrowable_publisher_`. If that's unavailable (i.e., if the `Optional<>` is uninitialized), it means the publisher was moved away from the stream to en external entity (the stream is in the "follower, not master" mode), and publishing to this stream is temporarily disabled (unless, of course, for the entity that borrowed the publisher -- and whoever else it has passed this publisher too).
2. Once acquired, a borrowed publisher can be passed around by the user code; more borrowers could be created, etc., etc.
3. Obviously, internally, `borrowable_publisher_`, if set, is borrowed from the `owned_publlisher_`.
4. When the stream is being created, or when it is asked to reacquire its own publisher (because it is ordered to become the "master", or because it should, atomically, "regain" the master status and give away the only ~virgin~ publisher away), all it needs to do is re-create the `owned_publisher_`. Destroying the optional instance of `owned_publlisher_` (obviously, nullifying the `borrowable_publisher_` first, and, obviously, the whole thing being done atomically, from within a mutex-locked section) would be the operation that notifies all previously issues borrowed publishers to terminate; and it will wait for them to terminate. Then, just re-creating this new "owned publisher" from scratch, pointed to `this`, will, when called from within a mutex-locked section, guarantee that no other entity can possibly have access to this publisher, enabling giving it away safely.

I've designed the above keeping the possibility of enabling the master-to-follower stream (or storage) flip in mind. Needless to say, we are not there yet. I just wanted to be ready.

I also separated stream and its data. Stream is the combination of "publishable" and "subscribable", while data is what can be synchronously iterated over. After some deliberation I ended up convincing myself that simple is the best strategy, and removed `.Empty()` and `.Size()` from the stream itself. The user would have to invoke `stream->Data()->Size()`. And, yes, while `Data()` could return a const reference, I've made it return a pointer because `->` is the trend now.

Oh, and I've removed some source (well, header) files that have been obsolete for a while now. Amen.

## Storage

Another piece which could eventually, in theory, go sideways one day with the present architecture is how we handle the master vs. follower storage.

Presently, the follower storage could become the master one, but not vice versa (the binary deprived of being the master storage only had one option: to turn its storage endpoints into 503-s and die peacefully). Thus, there were no mutexes at all: first there was a check the storage is the master one, then, if it is, the transaction was performed.

I've made the change to combine the `IsMaster` check and the actual transaction invocation to be within the same mutex-locked section now. This required changing the logic of the actual transaction invocation (our `SmartMutexLock` helped).

Also, if I remember correctly, the code before this PR contained a real issue of data being externally published into the stream of the type broader than `Transaction<>` when this stream is used as the base of the storage. I've cleaned up the architecture such that when a stream is owned by the master storage no external `.Publish()`-es into it may happen. Perhaps we'll end up extending the design so that `.Publish()` would be possible to be called on the storage itself, but only for the types that are not `Transaction<>`-s (a compile-time check, of course). This looks like a rather simple change, but I understand I already went too far in this refactoring, and decided to talk to you guys first before moving forward.

As a nice side effect, there is no longer any need in `StorageRole::{Master,Follower}` and `StreamDataAuthority::{Own,External}`. Both of these concepts are tied to `[underlying_stream_].IsMaster()`, which just confirms the `borrowable_publisher_` `Exists()`. Morever, that very `IsMaster()` call is actually a template `IsMaster<MutexLockSection MLS = NeedToLock>()` now, making it possible to perform atomic operations of the kind of "lock the mutex, now, if master, then X, otherwise Y" -- an example of which being the very `ReadWriteTransaction`, or a CQS command.

The above has one semantic "downside". I think it's good the way it is, but, of course, feedback is welcome. Basically, when creating a storage atop a stream the user need to specify if it should be a master storage or a following one, and just "whether an `Owned<>` or a `Borrowed<>` steam was passed in" doesn't appear to nail it. This becomes even worse with parameter-proxying constructor (the `template<typename... ARGS>` and `std::forward<ARGS>(args)...` one), because when the parameter is the file name, things get ugly (and nearly impossible to debug without cursing).

My solution for now is to explicitly require an owned or borrowed stream object -- effectively making the test syntax the following: `Storage storage(current::MakeOwned<typename Storage::stream_t>( ... stream construction params ... ))`. It looks good, although at times overly verbose to me. I'm most certainly open to suggestions on better semantics.

On a minor note, I've renamed `IsMaster` into `IsMaster{Storage,Stream}`, and made a few more name-clarifying changes for easier development and grep-ping.

## `Publish` and `UpdateHead`.

The user-facing syntax did not change, Internally, I've unified the code a bit, along two dimensions:

1. `Publish()` always takes two parameters. The first one is what to publish, and the second one is the timestamp. The timestamp can be `std::chrono::microseconds` or `current::time::DefaultTimeArgument`. The corresponding `ENABLE_IF`-s are in place. Same is true for `UpdateHead()`, which always needs one parameter. Reminder: This relates only to the internal implementation; the user-facing interfaces did not change, and the 2nd parameter to `.Publish()` (as well as the only parameter to `.UpdateHead()`) is not requires at all.
2. Both `Publish()` and `UpdateHead()` are now `template`-d, with the parameter being the `MutexLockStatus` one, defaulting to `NeedToLock`. This is done to enable appending to stream from a locked section. I've also invested into making sure publishing operations share one mutex (which includes the stream/storage/persister logic, as well as data authority flips).

Note: I've also added a `static_assert` (or a more narrow specification) to only allow passing the specific type to be published. I.e., if the stream is of type `Record`, and `Record` has a one-parameter constructor accepting an `int`, the user would still have to write `Publish(Record(42))` instead of `Publish(42)`. There was a painful part wrt `Variant<>`-s being the top-level types, but it's resolved now (albeit inefficiently).

Note 2: I've merged Grisha's change in, and now it's two template parameters. Warrants a chat.

Note 3: I've also made `MMPQ` a real "Publisher" now, which required me to sacrifice the `PublishIntoTheFuture` syntax. Now, MMPQ's `.Publish()` never moves the head pointer (as we don't know what timestamps can the next entries have), while `.UpdateHead()` is what moves the needle.

## Remaining items

Well. `Karl`'s test, while passing reliably on Travis, still seems to leak HTTP connections. It's reliably failing on my Linux with:

```
./.current/test --gtest_repeat=-1 --gtest_filter=Karl.ChangeDependenciesInClaire

# Fails after exactly 253 iterations with:

Karl.ChangeDependenciesInClaire : fails to resolve localhost
```

After trying this:
```
sudo sysctl net.ipv4.tcp_tw_reuse=1
sudo sysctl net.ipv4.tcp_tw_recycle=1
```

and even replacing `localhost` by `127.0.0.1`, it fails with `SocketCreateException`, as `::socket(AF_INET, SOCK_STREAM, IPPROTO_TCP)` returns a negative value -- on the same iteration.

I think it could be as simple as a long-playing attempt to reach a non-existing Karl from the test, from another thread; something another `Owned<>` could help with. We could research why do we create so many connections (debug output on the HTTP stack level would do the trick), and/or how do we recycle them better on our side. But the test no longer fails with random errors. Which is good. I wanted to make sure Travis is great again, largely for Ivan's enjoyment, and this goal is accomplished. To research further (`TODO(dkorolev)`, of course, it's on me), the magic command while running the above single test repeatedly is `lsof -Pan -p $(pgrep test) -i | wc -l`.

Also, perhaps now (after this PR is merged) is a good time to talk about mass renaming? We keep postponing the renaming of all the directories into lowercase; given we'll need to tweak RS' code after this PR, we might as well lowercase the dirs now, what do you think? Also, as our Storage is `Storage`, perhaps our Stream can be `Stream`, not `Sherlock`? :-) The name `StreamData` is also something I don't like. And, on top of the above, we have `Persister` of two different meanings: stream persister and storage persister. And this is not to mention that `Storage/` contains two header files that begin with `transaction_`. I don't believe the above list is exhaustive, but if we take care of the above, I'd say about 90% of my what bugs me will be gone, so why not do it?

* * *

More random notes from my scratchpad.

`TODO(dkorolev)`: We should eventually enable `Borrowed<BASE_T>` to be constructible from `{Owned|Borrowed}<DERIVED_T>`. A prime example would be some `Subscribable<T>`. We're not there yet though. -- D.K.

`TODO(dkorolev)`: When creating a `BorrowedWithCallback<T>` for an object that is already being destructed, should the callback be called right away? Should it be only called once (I assume so -- D.K.)? Now would be the right time to have this chat.

Tweaked the docus.
`CurrentHashFunction` -> `GenericHashFunction`.

Master flip for stream, then storage.

Add tests for no throws, even when spinning up new borrowers from an already-deleting owned.

Disable copying of the `fields` object of the DB (if easy).

Cast owned/borrowed to base type, as smart pointers do.

What does `GracefulShutdown` really do though? :-) Shall it be retired as the proper master flip logic is in place?

Add (back) the custom publisher for the stream that's now part of the storage. (storage->PublisherUsed()` is a cheat, after all.)

Smart dependencies makefile into current (test based on subdir file modifications, such as `Storage/persistence`)?

Rename Sherlock into Stream.

ReadWrite transactions might as well just borrow the publisher. This would enable us moving forward with `.Detach()`-ing them.

File and dir renaming

`HTTPResponseCode.` -> `HTTPResponseCode::`  (because we're still C++, even though it may look like Java now.)

Fix `./.current/run --scenario=stream_replication` in `examples/Benchmark/generic`.

Make use of `BorrowedOfGuaranteedLifetime<>`.

Make a note (and/or redesign) the HTTP handler registration for `*stream` -- it's unsafe, after all, although our usages look good to me.

* * *

Thanks,
Dima

mzhurovich commented on Blocks/HTTP/impl/posix_server.h:
```diff
@@ -346,18 +346,24 @@ class HTTPServerPOSIX final {
+                                                               URLPathArgs& output_url_args) const {
+    std::lock_guard<std::mutex> lock(mutex_);
+
     // Just `return` is safe. Uninitialized `output_handler` would be interpreted as "no handler found".
```
Fix the comment?

dkorolev replied:
:+1: 1f572f6

mzhurovich commented on Blocks/MMQ/test.cc:
```diff
@@ -282,8 +299,9 @@ TEST(InMemoryMQ, MMPQAllowsTimeExplicitlyGoingBack) {
 
   // Publish "two", "three", and "four". The first one to get published is "three", but, as it's published
   // into the future, it won't get processed until "four" is. Note the index of "two" is `3`, not `2`.
```
Fix the comment as `PublishIntoTheFuture` is deprecated? Not sure how to phrase it better, but it should mention `UpdateHead` imho.

dkorolev replied:
:+1: 11955dd

sompylasar commented on .gitignore:
```diff
@@ -1,54 +1,57 @@
+.idea
+CMakeLists.txt
+
+# Let's not even talk about it.
```
What a hater...

sompylasar commented on Blocks/Persistence/file.h:
```diff
@@ -393,31 +382,28 @@ class FilePersister {
-          begin_(begin),
-          end_(end),
-          begin_offset_(begin_offset) {}
+    IterableRangeImpl(current::Borrowed<FilePersisterImpl> file_persister_impl,
```
Just `Borrowed` above, `current::Borrowed` here?

dkorolev replied:
:+1: e9785a9

sompylasar commented on Blocks/Persistence/file.h:
```diff
@@ -426,117 +412,125 @@ class FilePersister {
+    end_t iterator = file_persister_impl_->end_.load();
+    const auto timestamp = current::time::GetTimestampFromLockedSection(std::forward<TIMESTAMP>(provided_timestamp));
     if (!(timestamp > iterator.head)) {
+#ifdef CURRENT_BUILD_WITH_PARANOIC_RUNTIME_CHECKS
```
The statement disabled by the `CURRENT_BUILD_WITH_PARANOIC_RUNTIME_CHECKS` flag is just output, not a check (the check and the throw are built in anyway); the name of the flag is misleading.

dkorolev replied:
True, but this symbol is used in other places too. Let's keep it as is?

sompylasar replied:
OK fine, let's.

sompylasar commented on Blocks/Persistence/file.h:
```diff
@@ -426,117 +412,125 @@ class FilePersister {
     }
 
     iterator.last_entry_us = iterator.head = timestamp;
     const auto current = idxts_t(iterator.next_index, iterator.last_entry_us);
```
`current` is confusing/ambiguous here; suggest renaming to `current_index` (or just `index`, because `entry`, not `current_entry`).

dkorolev replied:
:+1: e9785a9

sompylasar commented on Blocks/Persistence/memory.h:
```diff
@@ -24,10 +24,16 @@ SOFTWARE.
+// and accesses this "deque" by indexes from under a mutex when iterating over the entries.
 // Iterators never outlive the persister.
 
+// NOTE(dkorolev): I've took the liberty to use a dedicated `std::mutex` in `MemoryPersister`.
```
`I've taken` or `I took`.

dkorolev replied:
:+1: e9785a9

sompylasar commented on Bricks/net/http/test.cc:
```diff
@@ -422,6 +422,8 @@ TEST(PosixHTTPServerTest, ChunkedLargeBodyManyChunks) {
 }
 
+// This test sometimes fails on Apple on Travis. -- D.K.
+#ifndef CURRENT_APPLE
```
Maybe replace the disabled tests with something that will print that they were disabled, to make this obvious and explicit?

dkorolev replied:
:+1: e9785a9

sompylasar commented on Bricks/net/tcp/impl/posix.h:
```diff
@@ -509,20 +509,22 @@ struct Deleter {
   } else if (retval) {
     freeaddrinfo(result);
-    CURRENT_THROW(SocketResolveAddressException(host + (serv.empty() ? "" : " " + serv) + " " + gai_strerror(retval)));
+    CURRENT_THROW(SocketResolveAddressException(host_and_service + gai_strerror(retval)));
```
The space between the `host_and_service` and the error message from `gai_strerror` was lost here, so the error message will be glued with the `host_and_service`, example: `localhost:8080Some error message.` instead of `localhost:8080 Some error message.`

dkorolev replied:
:+1: e9785a9

sompylasar commented on Bricks/net/tcp/impl/posix.h:
```diff
@@ -509,20 +509,22 @@ struct Deleter {
   hints.ai_socktype = SOCK_STREAM;
   hints.ai_protocol = IPPROTO_TCP;
   const int retval = ::getaddrinfo(host.c_str(), serv.c_str(), &hints, &result);
+  addrinfo_t safe_result(result);
```
Is this OK to initialize `addrinfo_t safe_result` with `nullptr`?

dkorolev replied:
Made it safe in e9785a9

sompylasar commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,493 @@
+
+// Helper to prohibit copying `Owned`.
+struct ConstructOwned {};
+struct ConstuctBorrowable {};
```
`ConstuctBorrowable` -> `ConstructBorrowable`

dkorolev replied:
:+1: e9785a9

sompylasar commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,493 @@
+
+  // Smart-pointer-style accessors.
+  // NOT THREAD-SAFE, POTENTIALLY ALLOWS NON-EXCLUSIVE USE TO THE UNDERLYING OBJECT.
+  // THREAD-SAFE. NEVER THROWS.
```
```
// NOT THREAD-SAFE, POTENTIALLY ALLOWS NON-EXCLUSIVE USE TO THE UNDERLYING OBJECT.
// THREAD-SAFE. NEVER THROWS.
```
Contradicting lines?

dkorolev replied:
:+1: e9785a9

sompylasar commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,493 @@
+
+  // Exclusive accessors.
+  // THREAD-SAFE WITH RESPECT TO MAKING SURE UNDERLYING OBJECT IS ACCESSED EXCLUSIVELY BY THE PASSED IN FUNCTOR.
+  // THREAD-SAFE. NEVER THROWS.
```
```
// THREAD-SAFE WITH RESPECT TO MAKING SURE UNDERLYING OBJECT IS ACCESSED EXCLUSIVELY BY THE PASSED IN FUNCTOR.
// THREAD-SAFE. NEVER THROWS.
```
`THREAD-SAFE` twice?

dkorolev replied:
:+1: e9785a9

sompylasar commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,493 @@
+  Owned() : impl_t(), base_t(impl::ConstructOwned(), *impl_t::movable_instance_) {}
+
+#ifndef CURRENT_OWNED_BORROWED_EXPLICIT_ONLY_CONSTRUCTION
+  // For the end users for Current, we allow "manual" `Owned` construction via `ARGS...`.
```
`For the end users for Current` -> `For the end users of Current`

dkorolev replied:
:+1: e9785a9

sompylasar commented on Karl/test.cc:
```diff
@@ -744,12 +757,18 @@ TEST(Karl, ChangeKarlWhichClaireReportsTo) {
   }
 }
 
+// NOTE(@dkorolev): On my Ubuntu 16.04, when run as:
```
`NOTE(@dkorolev)` -> `NOTE(dkorolev)` for consistency, or vice versa.

dkorolev replied:
:+1: e9785a9

mzhurovich commented on Blocks/MMQ/mmpq.h:
```diff
@@ -70,83 +70,39 @@ class MMPQ {
-    if (!(timestamp > last_idx_ts_.us)) {
-      CURRENT_THROW(ss::InconsistentTimestampException(last_idx_ts_.us + std::chrono::microseconds(1), timestamp));
-    }
+    const auto us = current::time::GetTimestampFromLockedSection(std::forward<TIMESTAMP>(timestamp));
```
I've just looked what `GetTimestampFromLockedSection` is:
```
inline std::chrono::microseconds GetTimestampFromLockedSection(DefaultTimeArgument) { return Now(); }
inline std::chrono::microseconds GetTimestampFromLockedSection(std::chrono::microseconds us) { return us; }
```

Don't you think that:
1) The name of the function brings actually more confusion, as de-facto it's just a simple wrapper?
2) All this rvalue/forward syntax is an overkill and just hurts the readability?

dkorolev replied:
Good point. Renames and removed `forward`-s and extended references, in c977628.

mzhurovich commented on Blocks/MMQ/mmpq.h:
```diff
@@ -70,83 +70,39 @@ class MMPQ {
+  MMPQImpl(const MMPQImpl&) = delete;
+  MMPQImpl(MMPQImpl&&) = delete;
+  void operator=(const MMPQImpl&) = delete;
+  void operator=(MMPQImpl&&) = delete;
```
Not move-constructible?

dkorolev replied:
Yes. My line of thought was that it'd eventually be `Owned<>`, which is both more universal usage-wise and less code maintenance-wise. What do you think?

mzhurovich replied:
Okay :)

mzhurovich commented on Blocks/MMQ/test.cc:
```diff
@@ -338,28 +361,28 @@ TEST(InMemoryMQ, MMPQSupportsUpdateHead) {
   EXPECT_EQ("three @ 3, seven @ 7", current::strings::Join(c.messages_by_timestamps_, ", "));
 
   // Can't publish anything prior to "seven", as HEAD is already at `7`.
```
`Can't UpdateHead into the past`?

dkorolev replied:
:+1: 88cb4e7

mzhurovich commented on Blocks/MMQ/test.cc:
```diff
@@ -382,6 +405,7 @@ TEST(InMemoryMQ, MMPQSupportsUpdateHead) {
   // Finally, publish "joker", in the "present" which is way ahead in time compared to the third one left, the "jack".
   // This publish into "present" would force the consumer to process both the outstanding "jack" and the "joker" itself.
```
`Publish` doesn't force anything anymore.

dkorolev replied:
:+1: 88cb4e7

mzhurovich commented on Blocks/Persistence/file.h:
```diff
@@ -302,21 +317,16 @@ class FilePersister {
-        CURRENT_THROW(
-            PersistenceFileNoLongerAvailable(file_persister_impl_.ObjectAccessorDespitePossiblyDestructing().filename));
-      }
+    void operator++() {
```
Why `void`?

dkorolev replied:
Hmm, why not? :-) I'm uncertain whether it should return `bool` or `*this` to begin with.

dkorolev replied:
My bad, fixed.

mzhurovich commented on Blocks/Persistence/file.h:
```diff
@@ -426,117 +425,128 @@ class FilePersister {
+    file_persister_impl_->record_timestamp_.push_back(timestamp);
+
+    // Explicit `ConstructAsNecessary` is essential, otherwise the `Variant`'s case
+    // would be serialized in an unwrapped way when passed directly
```
Period at the end of the sentence.

dkorolev replied:
:+1: 88cb4e7

mzhurovich commented on Blocks/Persistence/file.h:
```diff
@@ -129,48 +146,49 @@ class FilePersister {
     // http://stackoverflow.com/questions/29824570/segfault-in-stdatomic-load/29824840#29824840
-    // std::atomic<end_t> end;
-    current::atomic_that_works<end_t> end;
+    // std::atomic<end_t> end_;
```
Add a `TODO` for conditional compilation? Proper atomics may boost the performance.

sompylasar replied:
Btw, the stackoverflow reference and the conditional compilation should likely be in the file where `current::atomic_that_works` is defined, selecting which impl to use: standard via typedef, or custom. Maybe move/rename to `current::sync::atomic` to co-locate sync primitives in one place.

dkorolev replied:
I suggest we stay conservative here. This part is to be retired soon regardless -- as we'd be moving to C++14, I assume.

Thoughts?

mzhurovich replied:
How C++14 would help us here?

mzhurovich commented on Blocks/Persistence/file.h:
```diff
@@ -426,117 +425,128 @@ class FilePersister {
+  }
+
+  template <current::locks::MutexLockStatus MLS>
+  bool PersisterEmptyImpl() const noexcept {
```
Strictly speaking, using `noexcept` here and below is wrong, as our atomic's `load()` constructs a `lock_guard` and may throw.

dkorolev replied:
:+1: 88cb4e7

dkorolev replied:
:+1: 88cb4e79105d908db139baac448bb988bfd0982c

mzhurovich commented on Blocks/Persistence/memory.h:
```diff
@@ -78,168 +87,158 @@ class MemoryPersister {
-    if (!container_->entries.empty()) {
-      return head_optidxts_t(container_->head, container_->entries.size() - 1, container_->entries.back().first);
+  template <current::locks::MutexLockStatus MLS>
+  head_optidxts_t PersisterHeadAndLastPublishedIndexAndTimestampImpl() const noexcept {
```
Same for `noexcept` here and below.

dkorolev replied:
:+1: 88cb4e7

mzhurovich commented on Blocks/Persistence/memory.h:
```diff
@@ -277,21 +276,21 @@ class MemoryPersister {
-      return Iterate<IM>(index_range.first, index_range.second);
-    } else {  // No entries found in the given range.
+      return PersisterIterateImpl<MLS, IM>(index_range.first, index_range.second);
+    } else {  // No entries_ found in the given range.
```
s/entries_/entries/

dkorolev replied:
:+1: 88cb4e7

mzhurovich commented on Blocks/Persistence/test.cc:
```diff
@@ -741,7 +755,8 @@ TEST(PersistenceLayer, FileSafeVsUnsafeIterators) {
     EXPECT_NO_THROW(*impl.Iterate().begin());
-    EXPECT_NO_THROW(*(++impl.Iterate().begin()));
+    // DIMA DIMA DIMA
+    // EXPECT_NO_THROW(*(++(impl.Iterate().begin())));
```
?

dkorolev replied:
Ah, I see what's wrong with `operator++` returning `void`. Fixed!

mzhurovich commented on Blocks/Persistence/test.cc:
```diff
@@ -751,11 +766,16 @@ TEST(PersistenceLayer, FileSafeVsUnsafeIterators) {
-    EXPECT_NO_THROW(*impl.Iterate<current::ss::IterationMode::Unsafe>().begin());
-    EXPECT_NO_THROW(*(++impl.Iterate<current::ss::IterationMode::Unsafe>().begin()));
+    // DIMA DIMA DIMA
+    // EXPECT_THROW(*(++impl.Iterate().begin()), current::ss::InconsistentIndexException);
```
?

dkorolev replied:
:+1: 88cb4e7

mzhurovich commented on Blocks/Persistence/test.cc:
```diff
@@ -770,10 +790,13 @@ TEST(PersistenceLayer, FileSafeVsUnsafeIterators) {
     // Check that "safe" iterators throw the MalformedEntryException while "unsafe" ones don't.
-    EXPECT_THROW(*(++(impl.Iterate().begin())), current::persistence::MalformedEntryException);
+    // DIMA DIMA DIMA
+    // EXPECT_THROW(*(++(impl.Iterate().begin())), current::persistence::MalformedEntryException);
```
?

dkorolev replied:
:+1: 88cb4e7

dkorolev replied:
Oh, I missed this one. Fixed in the next commit regardless.

mzhurovich commented on Bricks/net/http/test.cc:
```diff
@@ -673,6 +679,7 @@ TEST(PosixHTTPServerTest, ChunkedBoundaryCases) {
 }
 #endif  // CURRENT_HTTP_DATA_JOURNAL_ENABLED
 
+#ifndef CURRENT_APPLE  // This test is flaky on OS X.
```
Add `DISABLED_`?

dkorolev replied:
:+1: , in the previous commit.

mzhurovich commented on Bricks/net/tcp/impl/posix.h:
```diff
@@ -517,12 +518,13 @@ inline addrinfo_t GetAddrInfo(const std::string& host, const std::string& serv =
   }
-  return addrinfo_t(result);
+  addrinfo_t safe_result(result);
+  return safe_result;
```
How is it safer?

dkorolev replied:
"... сказала монашка надевая кондом на свечку"

Fixed in 88cb4e79105d908db139baac448bb988bfd0982c

mzhurovich commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,492 @@
+/*******************************************************************************
+The MIT License (MIT)
+
+Copyright (c) 2016 Dmitry "Dima" Korolev, <dmitry.korolev@gmail.com>.
```
2017?

dkorolev replied:
Well, it was 2016 when I `git mv`-ed the file. I prefer to stick with the philosophy of keeping the timestamp of the original code.

mzhurovich replied:
👍

mzhurovich commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,492 @@
+  // Return the number of registered borrower users besides the master one.
+  // For unit-testing purposes mostly, but may end up useful. -- D.K.
+  // THREAD-SAFE. NEVER THROWS.
+  size_t NumberOfActiveBorrowers() const noexcept {
```
Constructing a `lock_guard` may throw.

dkorolev replied:
:+1: 88cb4e79105d908db139baac448bb988bfd0982c

mzhurovich commented on Bricks/template/typelist.h:
```diff
@@ -75,7 +75,8 @@ struct TypeListContains;
 struct TypeListContains<TypeListImpl<TS...>, T> {
-  enum { value = std::is_base_of<DuplicateTypeInTypeList<T>, ConfirmTypeListContainsNoDuplicates<TS...>>::value };
+  constexpr static bool value =
+      std::is_base_of<DuplicateTypeInTypeList<T>, ConfirmTypeListContainsNoDuplicates<TS...>>::value;
```
👍

dkorolev commented:
@sompylasar -- I've changed one test into `DISABLED_` per your comment (1195d95). Re. the other cases your `pcregrep` points to, I'd rather keep them the way they are. The `#if[n?]def` guards there are not platform-related but logic-related; and this is not to mention the "%d DISABLED TESTS" message in the output of Travis would look unnecessarily alarming.

sompylasar commented:
>  this is not to mention the "%d DISABLED TESTS" message in the output of Travis would look unnecessarily alarming.

This looks like hiding the truth to me. The tests are disabled where they could be enabled. This is alarming.

dkorolev commented:
> This looks like hiding the truth to me. The tests are disabled where they could be enabled. This is alarming.

The tests for Mac are disabled on Linux. Is that hiding the truth or just reducing bloat?

sompylasar commented:
> The tests for Mac are disabled on Linux. Is that hiding the truth or just reducing bloat?

If they are the tests for a single platform only, they'd better be named accordingly, explicitly. E.g. `CanNotBindTwoSocketsToTheSamePortSimultaneously` -> `CanNotBindTwoSocketsToTheSamePortSimultaneouslyOnWindows`.

The following, if not displayed as disabled, give false assumptions that this functionality is tested in CI:
```
./FnCAS/test.cc:#ifdef FNCAS_JIT_COMPILED
// This test is here mostly to have the easily locatable autogenerated files in `/tmp` (on Linux).
TEST(FnCAS, TrivialCompiledFunctions) {
```
```
./FnCAS/test.cc:#ifdef FNCAS_JIT_COMPILED
TEST(FnCAS, CompiledGradientsWrapper) {
```
```
./FnCAS/test.cc:#ifdef FNCAS_JIT_COMPILED
TEST(FnCAS, OptimizationOfAStaticFunctionWithJIT) {
```
```
./FnCAS/test.cc:#ifdef FNCAS_JIT_COMPILED
TEST(FnCAS, OptimizationOfAPolynomialMemberFunctionWithJIT) {
```
```
./Karl/test.cc:#ifndef CURRENT_CI
TEST(Karl, DeregisterWithNginx) {
```
```
./Karl/test.cc:#ifndef CURRENT_CI
TEST(Karl, DisconnectedByTimoutWithNginx) {
```
```
./TypeSystem/test.cc:#ifndef VARIANT_CHECKS_AT_RUNTIME_INSTEAD_OF_COMPILE_TIME
// Nested variants are only allowed if variant type checking is compile-time, not runtime.
TEST(TypeSystemTest, NestedVariants)
```

mzhurovich commented on Blocks/SS/persister.h:
```diff
@@ -44,78 +49,81 @@ struct GenericEntryPersister : GenericPersister {};
+  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock,
+            typename E,
+            class = ENABLE_IF<CanPublish<current::decay<E>, ENTRY>::value>>
+  IndexAndTimestamp Publish(E&& e, current::time::DefaultTimeArgument = current::time::DefaultTimeArgument()) {
```
s/IndexAndTimestamp/idxts_t/g?

mzhurovich commented on Blocks/SS/test.cc:
```diff
@@ -81,28 +81,21 @@ struct DemoEntryPublisherImpl {
   std::deque<DispatchDemoEntry> values;  // Use `deque` to avoid extra copies.
 
   template <current::locks::MutexLockStatus, typename US>
```
s/US/TIMESTAMP/g for consistency?

mzhurovich commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,492 @@
+
+// Helper to prohibit copying `Owned`.
+struct ConstructOwned {};
+struct ConstructBorrowable {};
```
Whydid you use `...Borrowable` here, not `Borrowed`?

mzhurovich commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,492 @@
+    MoveFrom(std::move(rhs), destruction_callback);
+  }
+
+  void InitializedMovedOwned(instance_t& actual_instance) {
```
s/InitializedMovedOwned/InitializeMovedOwned/ ?

mzhurovich commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,492 @@
+  void InternalUnRegister() {
+    if (key_) {
+      CURRENT_ASSERT(p_actual_instance_);
+      // `*this` is a valid (non - std::move()-d - away) `BorrowedWithCallback`.
```
This function is also called by `Borrower`, so looks like this is not always true.

mzhurovich commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,492 @@
+// 1) `const WeakBorrowed<T>&`, and
+// 2) a required callback, which will be called once the scope of the `Owned<T>` will be terminating.
+// Obviously, an instance of `BorrowedWithCallback` can't be moved natively, as the
+// destruction callback it uses must point to the respective owner object, not to the "old" one.
```
Could you elaborate pls? Callback is just a `std::function<void()>` and doesn't point to anything.

mzhurovich commented:
Btw, top-level `make test` is broken

```
In file included from everything.cc:63:
.././examples/Benchmark/Sherlock/load_test.cc:16:5: error: redefinition of 'main'
int main(int argc, char** argv) {
    ^
.././examples/Benchmark/HTTP/../../../3rdparty/gtest/gtest-main-with-dflags.h:40:5: note: previous definition is here
int main(int argc, char** argv) {
    ^
In file included from everything.cc:63:
.././examples/Benchmark/Sherlock/load_test.cc:20:17: error: no matching conversion for functional-style cast from 'std::string' (aka 'basic_string<char,
      char_traits<char>, allocator<char> >') to 'current::sherlock::Stream<Entry, current::persistence::File>' (aka 'Stream<CSI_Entry< ::crnt::r::DF>,
      current::persistence::File>')
  auto stream = current::sherlock::Stream<Entry, current::persistence::File>(FLAGS_file);
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

mzhurovich commented on Bricks/sync/test.cc:
```diff
@@ -22,39 +22,95 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-  EXPECT_EQ(1, *x);
-  EXPECT_TRUE(static_cast<bool>(x));
+// Test using `current::Owned<>` as a `shared_ptr<>`.
+TEST(Borrowable, Owned) {
```
s/Borrowable/OwnedBorrowed/?

mzhurovich commented on Bricks/sync/test.cc:
```diff
@@ -65,51 +121,51 @@ TEST(ScopeOwned, MoveConstructScopeOwnedByMe) {
+TEST(Borrowable, BorrowedWithCallbackOutlivingTheOwner) {
   // Declare the variable to work with, and a reference wrapper for it, outside the scope of the test.
-  // This way the resulting, final value of this variable can be tested as the `ScopeOwned*` part is all done.
+  // This way the resulting, final value of this variable can be tested as the `Borrowable*` part is all done.
```
`Borrowable*` -> `Owned/Borrowed`?

mzhurovich commented on Bricks/sync/test.cc:
```diff
@@ -161,8 +219,8 @@ TEST(ScopeOwned, ScopeOwnedBySomeoneElseOutlivingTheOwner) {
 // Same as the above test, but do not use an external termination primitive,
-// but rely on polling the availability of `ScopeOwnedBySomeoneElse<>` itseld.
-TEST(ScopeOwned, UseInternalIsDestructingGetter) {
+// but rely on polling the availability of `current::BorrowedWithCallback<>` itseld.
```
s/itseld/itself/

mzhurovich commented on Bricks/sync/waitable_atomic.h:
```diff
@@ -329,7 +329,7 @@ class WaitableAtomicImpl {
 
    protected:
-    bool destructing_ = false;
+    volatile bool destructing_ = false;
```
atomic_bool?

mzhurovich commented on Sherlock/sherlock.h:
```diff
@@ -87,208 +91,185 @@ constexpr const char* kDefaultTopLevelName = "TopLevelTransaction";
-    // Order of destruction in `http_subscriptions` does matter - scopes should be deleted first -- M.Z.
-    for (auto& it : own_data_.ObjectAccessorDespitePossiblyDestructing().http_subscriptions) {
-      it.second.first = nullptr;
+  // `BorrowPublisher()`: The called can count on the fact that if the call succeeded,
```
s/called/caller/

mzhurovich commented on Sherlock/sherlock.h:
```diff
@@ -87,208 +91,185 @@ constexpr const char* kDefaultTopLevelName = "TopLevelTransaction";
+  // `Publisher()`: The caller assumes full responsibility for making sure the underlying stream
+  // is not destroyed while it is being published to.
+  // CAN THROW `PublisherNotAvailableException`.
+  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock>
```
I have a gut feeling that exposing `template <current::locks::MutexLockStatus MLS>` functions to the end user is not good :(

dkorolev replied:
Spoke of this with @grixa on a late night call. Will refactor soon.

mzhurovich commented on Sherlock/sherlock.h:
```diff
@@ -300,36 +281,38 @@ class StreamImpl {
-        // The constructor has not completed successfully. The thread was not started, and `data_` is garbage.
+        // The constructor has not completed successfully. The thread was not started, and `impl_` is garbage.
         if (done_callback_) {
           // TODO(dkorolev): Fix this ownership issue.
```
Is it fixed?

mzhurovich commented on Sherlock/sherlock.h:
```diff
@@ -417,178 +398,192 @@ class StreamImpl {
-        return;
-      }
+    // First, ensure we're not in the destruction mode, and atomically claim we should not be
+    // without destructing this HTTP subscription, assuming we do follow up with creating one.
```
I don't understand what `atomically claim we should not be without destructing this HTTP subscription` means :(

mzhurovich commented on Sherlock/sherlock.h:
```diff
@@ -417,178 +398,192 @@ class StreamImpl {
+    // First, ensure we're not in the destruction mode, and atomically claim we should not be
+    // without destructing this HTTP subscription, assuming we do follow up with creating one.
+    const Borrowed<impl_t> borrowed_impl(impl_);
+    if (!borrowed_impl) {
```
Constructing an object to perform the check looks a bit weird, shouldn't `IsValid()` do the trick?

dkorolev replied:
`IsValid` would do (and so would just `if (impl_)`, but the subscriber should borrow the stream data to work with it and to make sure it doesn't disappear in the meantime.

mzhurovich commented on port.h:
```diff
@@ -225,6 +225,10 @@ using namespace enable_strptime_on_windows;
   fprintf(stderr, "Current assertion failed:\n\t%s\n\t%s : %d\n", text, file, line);
+#ifdef CURRENT_BUILD_WITH_PARANOIC_RUNTIME_CHECKS
+  volatile int* fffuuuuu = nullptr;
+  *fffuuuuu = 42;
```
What is that?

dkorolev replied:
A canonical way to throw a SEGFAULT.

sompylasar replied:
Maybe name the variable `segfault` to be clear?

dkorolev replied:
How about `you_shalt_segfault`?daba665

dkorolev replied:
How about `you_shalt_segfault`? daba665

mzhurovich commented on Storage/storage.h:
```diff
@@ -249,80 +274,93 @@ class GenericStorageImpl {
+    persister_.BecomeMasterStorage();
   }
 
   void GracefulShutdown() { transaction_policy_.GracefulShutdown(); }
```
As you removed the lock_guards from the transaction policy, the user has no more control over the actual moment when the storage shuts down after the call to `GracefulShutdown()` :-/

dkorolev replied:
Let's take this one offline. The value of `GracefulShutdown` is vague for me as of right now, as it appear to not make sense outside the usecases of "prepare to make the storage read-only (perhaps temporarily because a master flip is being prepared") and "stop serving storage data via HTTP as the storage is being destructed".

mzhurovich commented on Storage/persister/sherlock.h:
```diff
@@ -120,66 +132,100 @@ class SherlockStreamPersisterImpl {
+  // Important: The publishing mutex of the respective stream must be unlocked!
+  void TerminateStreamSubscriptionFromLockedSection() { subscriber_scope_ = nullptr; }
+
+  void PushTimestampToFromLockedSection(std::chrono::microseconds timestamp) {
```
`SetLastAppliedTimestampFromLockedSection` is longer but cold be less ambiguous imho

mzhurovich commented on Storage/persister/sherlock.h:
```diff
@@ -120,66 +132,100 @@ class SherlockStreamPersisterImpl {
-  std::unique_ptr<SherlockSubscriber> subscriber_;
+
+  std::mutex& stream_publishing_mutex_ref_;                            // == `stream_->Impl()->publishing_mutex`.
+  Borrowed<stream_t> stream_;                                          // TODO(dkorolev): Rename to just `stream_`.
```
`TODO` looks obsolete.

mzhurovich commented on Storage/persister/sherlock.h:
```diff
@@ -120,66 +132,100 @@ class SherlockStreamPersisterImpl {
+      TerminateStreamSubscriptionFromLockedSection();
+      std::lock_guard<std::mutex> lock(stream_publishing_mutex_ref_);
+      publisher_used_ = nullptr;
+      publisher_used_ = stream_->template BecomeFollowingStream<current::locks::MutexLockStatus::AlreadyLocked>();
```
Why `BecomeMasterStorage` invokes `BecomeFollowingStream` for its stream?

dkorolev replied:
Very good question. TL;DR: It's a naming issue.

`BecomeFollowingStream` on the stream is, in essence, just releasing this stream's publisher to an external authority. I've debated a lot re. this name, and concluded, perhaps wrongly, "become follower stream" is the shortest way to convey the message.

In a way, storage becoming a master storage _impllies_ its underlying stream becomes a following stream -- as this underlying stream's publisher is being released to be exclusively owned by the storage itself.

After some thinking now, it still looks clean and simple (although possibly controversial at the first glance) to me. Perhaps we could add a comment here. Suggestions warmly welcome.

grixa commented on Sherlock/pubsub.h:
```diff
@@ -226,13 +226,13 @@ inline ParsedHTTPRequestParams ParsePubSubHTTPRequest(const Request& r) {
 
   PubSubHTTPEndpointImpl(const std::string& subscription_id,
-                         ScopeOwned<stream_data_t>& data,
+                         current::Borrowed<impl_t> data,
```
Just `Borrowed<impl_t>` is enough, isn't it?

grixa commented on Sherlock/sherlock.h:
```diff
@@ -300,36 +281,38 @@ class StreamImpl {
 
    public:
-    SubscriberThreadInstance(ScopeOwned<stream_data_t>& data,
+    SubscriberThreadInstance(current::Borrowed<impl_t> impl,
```
Same here - `current::` is redundant

grixa commented on Sherlock/sherlock.h:
```diff
@@ -417,178 +398,192 @@ class StreamImpl {
     using subscriber_thread_t = SubscriberThreadInstance<TYPE_SUBSCRIBED_TO, F>;
 
-    SubscriberScope(ScopeOwned<stream_data_t>& data,
+    SubscriberScope(current::Borrowed<impl_t> impl,
```
redundant `current::`. Or is it on purpose?

dkorolev commented:
@sompylasar , re. your test renaming suggestions:
1) See 6b5bbff.
2) your original `OnWindows` comment would in reality be `ExceptWindows`, which is kind of against the point.
3) the `NestedVariants` one is in fact testing the functionality that had to be enabled by that very `#define`-d symbol in order to be present. I'm open for the conversation to retire that, but, so far, why don't we just keep it as is?

dkorolev commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,492 @@
+
+// `UniqueInstanceContainer<T>` must be a dedicated struct for `Owned<T>` to construct
+// the actual object before initializing the `WeakBorrowed<T>` associated with it.
+// `UniqueInstanceContainer<T>` is constructible, movable, and nothing else.
```
1) I'd be reluctant to inherit from `std::unique_prtr`.
2) In `Stream` and `Storage`, for the static `Create*(*)` "constructors", `friend class UniqueInstance` is the way to go.

grixa commented on Blocks/Persistence/file.h:
```diff
@@ -115,6 +119,19 @@ class IteratorOverFileOfPersistedEntries {
 };
 
+template <typename DESIRED, typename ACTUAL>
+struct ConstructAsNecessary {
```
May be we should call it, for instance, `(Wrap|Prepare)(To|Before)Serialize`?

dkorolev replied:
`MakeSureTheRightTypeIsSerialized` :-)

`git push` coming.

dkorolev commented:
@mzhurovich @grixa @sompylasar 

Thanks for the review! Addressed all the comments (in 1602c8ee9e8601c72335823d6c8cc1883aca067e and 454d9ab99b74d18fa14ae7a61784c44aeca2c88e 6b5bbffeb6dbabbbb4ac11f61d62580b09daa4fe respectively), and commented on those few where the "fix" was not so obvious.

Friendly reminder: this change goes into `alt-master`, not `master`, and thus we can allow ourselves some flexibility (to the best of my knowledge, all four of us have related changes that are blocked by this PR being open atm).

Also CC @nyospe: Nathan, look what we have! It should make fixing the flaky tests much easier.

Thanks,
Dima

grixa commented:
The whole idea (ideas) is LGTM and surely worth to be merged :) I will look through the changes more thoroughly later, but don't wait for me and merge now if everyone agrees.

dkorolev commented:
Thanks! @mzhurovich @sompylasar? Another good reason to merge this one is that GitHub loading all the comments takes a significant while :(

sompylasar commented:
> @sompylasar , re. your test renaming suggestions:
> 1) See 6b5bbff.
> 2) your original `OnWindows` comment would in reality be `ExceptWindows`, which is kind of against the point.
> 3) the `NestedVariants` one is in fact testing the functionality that had to be enabled by that very `#define`-d symbol in order to be present. I'm open for the conversation to retire that, but, so far, why don't we just keep it as is? 

Thanks! Keep as is, let it be.

> Thanks! @mzhurovich @sompylasar? Another good reason to merge this one is that GitHub loading all the comments takes a significant while :(

Get a faster internet 😝 kidding, LGTM anyway. I haven't looked too deep into it because I'm not too confident with what's happening here regarding the ownership patterns.

dkorolev commented:
> Get a faster internet.

A faster browser, I'd assume. Somehow in the 21st century, where we can plan seemingly realistic video games, rendering a mere ~7K lines diff at anything above 0.5FPS is considered tricky.

OH! I KNOW! IT'S BECAUSE BROWSERS! :trollface:

mzhurovich commented on Sherlock/sherlock.h:
```diff
@@ -436,8 +435,8 @@ class Stream final {
 
-    // First, ensure we're not in the destruction mode, and atomically claim we should not be
-    // without destructing this HTTP subscription, assuming we do follow up with creating one.
+    // First and foremost, save outselves the trouble of starting the subscriber
```
s/outselves/ourselves/

mzhurovich commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -74,7 +74,7 @@ namespace impl {
 struct ConstructOwned {};
 struct ConstructBorrowable {};
-struct ConstructBorrowableObjectViaMoveConstructor {};
+struct ConstructBorrowedObjectViaMoveConstructor {};
```
s/ConstructBorrowable/ConstructBorrowed/g as well?

mzhurovich commented on Storage/persister/sherlock.h:
```diff
@@ -71,44 +71,56 @@ class SherlockStreamPersisterImpl {
 
-  void PersistJournal(MutationJournal& journal) {
+  template <current::locks::MutexLockStatus MLS>
+  std::chrono::microseconds UpToDateUntilPersister() const {
```
Maybe just `LastAppliedTimestamp`? Why introduce different names for a single concept?

grixa commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,492 @@
+// Philosophy:
+// * The scope in which the object has been created is the master owner of this object; others are borrowers.
+// * Upon leaving the master scope, the synchronization primitive waits until all its borrowers terminate.
+// * Borrowers are under no obligation to terminate sooner or later, but they get signaled to wrap up.
```
As I understand, the latter is not true for the generally used `Borrower`, it is for the special` BorrowedWithCallback` case only.

dkorolev replied:
:+1: rephrased.

grixa commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,492 @@
+// * Borrowed<T> borrower(master);  // Or `Borrowed<T> another_borrower(borrower);`.
+//   Initializes `borrower`, which can be used same way as `master`.
+//   Casting it to `bool` (or just checking for `if (!borrower) { ... }`) would result in `false`
+//   if and only if the master scope (the scope of the `Owned<T>` is being terminated, and is currently
```
`...the master scope (the scope of the `Owned<T>`**)** is being...`

grixa commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,492 @@
+//   Creates `master`, which will live only within its creation scope.
+//   The basic usage of `master` is `master->MemberFunctionOfT(...)`, much like with C++'s `shared_ptr<T>`.
+//
+// * Borrowed<T> borrower(master);  // Or `Borrowed<T> another_borrower(borrower);`.
```
As for me, the inverted second part (`// Or Borrowed<T> borrower(another_borrower);`) looks a little more clear and symmetric.

dkorolev replied:
Very good point. Fixed.

grixa commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,492 @@
+    // Wait until all the borrowers have terminated.
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
+      condition_variable_.wait(lock, [this]() { return borrowers_.empty(); });
```
As it is a potential cause of deadlocks, could we print to `cerr` some warning every second (or two, or more) of waiting? At least, in debug or in `CURRENT_BUILD_WITH_PARANOIC_RUNTIME_CHECKS` mode.

dkorolev replied:
Great idea. I'm also hoping to eventually use the `BorrowedOfGuaranteedLifetime`.

Could you remind me next time we chat?

grixa commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,492 @@
+// * BorrowedWithCallback<T> borrower(master, [&signal]() { signal.SignalTermination(); });
+//   Initializes `borrower` passing it an external signal that will be likely `std::move()` it into a different thread.
+//   The lifetime of the `borrower` object will be strictly inside the lifetime of `master`.
+//   The destructor of `master` will wait until all the `borrower`-s have terminated.
```
`Borrowed` and `BorrowedWithCallback` is very similar, but their descriptions look very different at first sight. I suggest two solutions: 1) list all matching "features" twice (here and above) with the same words or 2) first mention here that "the `BorrowedWithCallback` is the same as the `Borrowed`, except...", and then list only differences (not move-constructible and has the callback).

grixa commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,492 @@
+
+// `WeakBorrowed<T>` is the base class for `Owned<T>`, `Borrowed<T>`, and `BorrowedWithCallback<T>`.
+template <typename T>
+class WeakBorrowed {
```
May be it is worth to mention this class in the implementation details above, because it is actually used outside this header, something like:
"The `WeakBorrowed<T>` is the base class for all of them, so it is used if the code doesn't depend on the object ownership details."

dkorolev replied:
Rephrased. Push coming after local tests pass.

dkorolev commented:
@grixa, thanks for the review! Fixed your and @mzhurovich's comments in 4910c125fd562a0bb0e90b2bb49881411602a100, PTAL.

grixa commented on Storage/storage.h:
```diff
@@ -191,54 +191,79 @@ using CURRENT_STORAGE_DEFAULT_PERSISTER_PARAM = persister::NoCustomPersisterPara
+  }
+
+  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock>
+  std::chrono::microseconds LastApliedTimestamp() const {
```
I believe it should be LastA**pp**liedTimestamp

dkorolev replied:
:+1: 1793c60

grixa commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,493 @@
+//   Creates `master`, which will live only within its creation scope.
+//   The basic usage of `master` is `master->MemberFunctionOfT(...)`, much like with C++'s `shared_ptr<T>`.
+//
+// * Borrowed<T> borrower(master);  // Or `Borrowed<T> borrower(another_another_borrower);`.
```
Recursive `another_`. Is it a typo or feature?

dkorolev replied:
:+1: 8107602

dkorolev replied:
:+1: , fixed in 8107602

grixa commented on Bricks/sync/owned_borrowed.h:
```diff
@@ -0,0 +1,493 @@
+class BorrowedOfGuaranteedLifetime;
+
+// `WeakBorrowed<T>` is the base class for `Owned<T>`, `Borrowed<T>`, and `BorrowedWithCallback<T>`,
+// so it is used if the code doesn't depend on the details of original object ownership.
```
Well, I meant to add this to the big description in the beginning of the file, but never mind.

dkorolev replied:
:innocent:
